cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringAgentService.java

```diff
@@ -335,39 +365,22 @@ private void publishStopProgram(String stopPayload) {
   }
 
   /**
-   * Fetch program status updates and filter if they are initiated by a tethered peer.
+   * Add notification to map if associated by a tethered peer
    */
-  private Map<String, TetheringProgramUpdate> getPeerProgramUpdates() {
-    String lastMessageId = lastProgramUpdateMessageId;
+  private void addPeeredNotification(Notification notification, Map<String, List<Notification>> peerToNotifications) {
+    Map<String, String> properties = notification.getProperties();
+    String programRunId = properties.get(ProgramOptionConstants.PROGRAM_RUN_ID);
     try {
-      Map<String, TetheringProgramUpdate> peerProgramUpdates = new HashMap<>();
-      try (CloseableIterator<Message> iterator = messageFetcher.fetch(NamespaceId.SYSTEM.getNamespace(),
-                                                                      programUpdateTopic,
-                                                                      programUpdateFetchSize,
-                                                                      lastProgramUpdateMessageId)) {
-        if (iterator.hasNext()) {
-          Message message = iterator.next();
-          Notification notification = message.decodePayload(r -> GSON.fromJson(r, Notification.class));
-          Map<String, String> properties = notification.getProperties();
-          String programRunId = properties.get(ProgramOptionConstants.PROGRAM_RUN_ID);
-          String programStatus = properties.get(ProgramOptionConstants.PROGRAM_STATUS);
-          try {
-            RunRecord runRecord = runRecordFetcher.getRunRecordMeta(GSON.fromJson(programRunId, ProgramRunId.class));
-            if (runRecord.getPeerName() != null) {
-              peerProgramUpdates.getOrDefault(runRecord.getPeerName(), new TetheringProgramUpdate())
-                .addUpdateRequest(programRunId, programStatus);
-            }
-          } catch (NotFoundException e) {
-            LOG.error("Unable to fetch runRecord for programRunId {}", programRunId, e);
-          }
-          lastProgramUpdateMessageId = message.getId();
-        }
+      RunRecord runRecord = runRecordFetcher.getRunRecordMeta(GSON.fromJson(programRunId, ProgramRunId.class));
+      if (runRecord.getPeerName() != null) {
+        peerToNotifications.getOrDefault(runRecord.getPeerName(), new ArrayList<>()).add(notification);
       }
-      return peerProgramUpdates;
-    } catch (Exception e) {
-      LOG.error("Exception when fetching program updates. Will retry again during next poll", e);
-      lastProgramUpdateMessageId = lastMessageId;
-      return new HashMap<>();
+    } catch (NotFoundException | IOException e) {
+      LOG.error("Unable to fetch runRecord for programRunId {}", programRunId, e);
```

Comment:
should let the exception propagate up instead of log and continue. The superclass will retry processing the messages again. Otherwise things like transient DB issues will cause data loss where notifications are skipped.

Reply:
fixed